hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/connector/ReadWithPartitionIteratorP.java

```diff
@@ -0,0 +1,329 @@
+                }
+                iteratorCursor.remove();
+            } while (iteratorCursor.advance());
+            return null;
+        };
+    }
+
+    public static ProcessorMetaSupplier readMap(String mapName) {
+        return readMap(mapName, DEFAULT_FETCH_SIZE);
+    }
+
+    public static ProcessorMetaSupplier readMap(String mapName, int fetchSize) {
+        return new LocalClusterMetaSupplier(fetchSize,
+                instance -> instance.getMap(mapName));
+    }
+
+    public static ProcessorMetaSupplier readMap(String mapName, ClientConfig clientConfig) {
+        return readMap(mapName, DEFAULT_FETCH_SIZE, clientConfig);
+    }
+
+    public static ProcessorMetaSupplier readMap(String mapName, int fetchSize, ClientConfig clientConfig) {
+        return new RemoteClusterMetaSupplier(fetchSize, clientConfig,
+                instance -> instance.getMap(mapName));
+    }
+
+    public static ProcessorMetaSupplier readCache(String cacheName) {
+        return readCache(cacheName, DEFAULT_FETCH_SIZE);
+    }
+
+    public static ProcessorMetaSupplier readCache(String cacheName, int fetchSize) {
+        return new LocalClusterMetaSupplier(fetchSize,
+                instance -> instance.getCacheManager().getCache(cacheName));
+    }
+
+    public static ProcessorMetaSupplier readCache(String cacheName, ClientConfig clientConfig) {
+        return readCache(cacheName, DEFAULT_FETCH_SIZE, clientConfig);
+    }
+
+    public static ProcessorMetaSupplier readCache(String cacheName, int fetchSize, ClientConfig clientConfig) {
+        return new RemoteClusterMetaSupplier(fetchSize, clientConfig,
+                instance -> instance.getCacheManager().getCache(cacheName));
+    }
+
+    @Override
+    public boolean complete() {
+        return emitCooperatively(outputTraverser);
+    }
+
+    @Override
+    public boolean isCooperative() {
+        return false;
+    }
+
+    private static List<Processor> getProcessors(int count, List<Integer> ownedPartitions,
+                                                 Function<Integer, Iterator<Map.Entry>> partitionToIterator) {
+        Map<Integer, List<Integer>> processorToPartitions =
+                range(0, ownedPartitions.size())
+                        .mapToObj(i -> entry(i, ownedPartitions.get(i)))
+                        .collect(groupingBy(e -> e.getKey() % count, mapping(Map.Entry::getValue, toList())));
+        range(0, count).forEach(processor -> processorToPartitions.computeIfAbsent(processor, x -> emptyList()));
+        return processorToPartitions
+                .values().stream()
+                .map(partitions -> !partitions.isEmpty()
+                        ? new ReadWithPartitionIteratorP(partitionToIterator, partitions)
+                        : new Processors.NoopP()
+                )
+                .collect(toList());
+    }
+
+    private static Iterator<Map.Entry> partitionIterator(DistributedObject distributedObject, int fetchSize, int partitionId) {
```

Comment:
would be nicer to replace this method with a `Function<DistributedObject, Iterator<Map.Entry>>` instead.